src/shared/mount-util.c

```diff
@@ -325,46 +332,67 @@ int bind_remount_recursive_with_mountinfo(
                 }
 
                 /* If we have no submounts to process anymore, we are done */
-                if (set_isempty(todo))
+                if (hashmap_isempty(todo))
                         return 0;
 
-                while ((x = set_steal_first(todo))) {
+                for (;;) {
+                        unsigned long flags;
+                        char *x = NULL;
+
+                        /* Take the first mount from our list of mounts to still process */
+                        flags = PTR_TO_ULONG(hashmap_steal_first_key_and_value(todo, (void**) &x));
+                        if (!x)
+                                break;
 
-                        r = set_consume(done, x);
+                        r = set_ensure_consume(&done, &path_hash_ops_free, x);
                         if (IN_SET(r, 0, -EEXIST))
-                                continue;
+                                continue; /* Already done */
                         if (r < 0)
                                 return r;
 
-                        /* Deal with mount points that are obstructed by a later mount */
-                        r = path_is_mount_point(x, NULL, 0);
-                        if (IN_SET(r, 0, -ENOENT))
-                                continue;
+                        /* Now, remount this with the new flags set, but exclude MS_RELATIME from it. (It's
+                         * the default anyway, thus redundant, and in userns we'll get an error if we try to
+                         * explicitly enable it) */
+                        r = mount_nofollow(NULL, x, NULL, ((flags & ~flags_mask)|MS_BIND|MS_REMOUNT|new_flags) & ~MS_RELATIME, NULL);
```

Comment:
If `MS_RELATIME` is ignored, shouldn't `MS_NOATIME`, `MS_NODIRATIME`, `MS_STRICTATIME`, `MS_SYNCHRONOUS` be too?

Reply:
so the four MS_XXXATIME flags are exclusive to each other, you can only have one of them set. On top of that you have to have one of them set. Now oftentimes none is specified when mounting things, in which case the kernel will turn on MS_RELATIME for you automatically, as that's the kernel-default these days.

Now to make things worse: in userns changing these flags is typically disallows, so if you set one of those four flags there the kernel will EPERM you.

So in essence the best approach is to never try to set MS_RELATIME (it's unnecessary after all, because the absence of the other 3 flags imply it), and to ignore it when comparing what is set (since even if we set it to off it will be set when coming back), which is what we do.

The other flags you mentioned aren't so weird. they are not magically added in if you don't mention them but only set on request.